File: afamar-frontend/src/types/workOrder.ts
// Work order payload — English snake_case field names matching backend API exactly.
// Source of truth: afamar-backend/app/schemas/work_order.py (WorkOrderBase / WorkOrderCreate / WorkOrderUpdate).

import type { FabricationDetail, MaterialInForm, PoolInForm } from './budget';
import type { FinancialBase } from './shared';

export interface ConvertOptionResponse {
  message: string;
  orderId: number;
  number: string;
}

export interface WorkOrderPayload extends FinancialBase {
  client_name: string | null;
  client_phone: string | null;
  client_email: string | null;
  client_address: string | null;
  /** Optional override — when the customer wants the work done at a
   *  non-default address (e.g. an architect with several project sites).
   *  `null` means "use the client's default address". */
  delivery_address_id: number | null;
  budget_id: number | null;
  material: string | null;
  material_price_m2: number;
  materials_data: string | null;
  color: string | null;
  thickness: string | null;
  finish: string | null;
  bacha: string | null;
  anafe: string | null;
  delivery_date: string | null;
  digital_signature: string | null;
  fabrication_details: string | null;
  budgeted_details: string | null;
  /** JSON-encoded array of sketch PAGES (wire format
   *  `[{ pagina_id, name, material?, dibujo: [...] }]`) — the page shape is
   *  what WorkOrders persist so the FICHA DE TALLER can render a per-page
   *  name/material label. Populated from the source budget on conversion
   *  (legacy flat lists) or from the editor. When the user edits the WO,
   *  the editor mutates this list and we re-serialise via
   *  `serializeSketchPages` in `buildPayload`. */
  sketch_elements: string | null;
  pool_id: number | null;
  pool_price: number;
  pool_currency: string;
  pool_image: string | null;
  pools_data: string | null;
  additional_works_data: string | null;
  design_observations: string | null;
  important_observations: string | null;
  notes: string | null;
  date: string | null;
  status?: string;
}

/**
 * Trimmed shape returned by GET /work-orders (list). The backend populates
 * client_name/phone/email/address from the related Client row.
 *
 * Multi-piece v3: the list response carries BOTH the legacy flat
 * `material` column AND `materials_data` (JSON string) so the table
 * can render the principal name even when the legacy column was
 * cleared during piece-driven editing.
 */
export interface WorkOrderListItem {
  id: number;
  number: string;
  status: string;
  origin?: string;
  client_id?: number | null;
  budget_id?: number | null;

  // Client
  client_name: string | null;
  client_phone: string | null;
  client_email: string | null;
  client_address: string | null;

  // Material — multiple sources, table picks the first non-empty:
  material: string | null;
  /** Parsed JSON array of multi-piece materials (v3 source of truth).
   *  Each entry is `{ name, length, width, quantity, price_m2, ... }`. */
  materials_data?: string | null;
  /** Parsed JSON array of pieces. Each piece carries
   *  `mainMaterial: { name, ... }` + `alternativeMaterials[]`. */
  pieces?: unknown[] | null;
  /** Legacy field (very old list responses). Treat as `material`. */
  items?: unknown[] | null;
  estimated_delivery_date?: string | null;

  // Money (FinancialBase sub set)
  currency: string;
  total: number;
  total_usd?: number;
  deposit_received: number;
  balance_due: number;

  // Dates
  delivery_date: string | null;
  date?: string | null;
  created_at?: string;
}

File: afamar-backend/app/schemas/work_order.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel

from app.schemas.base import BaseResponse


class WorkOrderBase(BaseModel):
    client_id: int | None = None
    client_name: str | None = None
    client_phone: str | None = None
    client_email: str | None = None
    client_address: str | None = None
    budget_id: int | None = None
    material: str | None = None
    material_price_m2: float = 0.0
    materials_data: str | None = None
    color: str | None = None
    thickness: str | None = None
    finish: str | None = None
    bacha: str | None = None
    anafe: str | None = None
    currency: str = "ARS"
    usd_rate: float = 1000.0
    subtotal: float = 0.0
    transport: float = 0.0
    installation: float = 0.0
    discount: float = 0.0
    discount_percentage: float = 0.0
    discount_fixed_amount: float = 0.0
    total: float = 0.0
    subtotal_usd: float = 0.0
    transport_usd: float = 0.0
    total_usd: float = 0.0
    deposit_received: float = 0.0
    deposit_currency: str = "ARS"
    deposit_usd: float = 0.0
    balance_due: float = 0.0
    balance_due_usd: float = 0.0
    payment_method: str | None = None
    installments: int = 1
    priority: str = "NORMAL"
    delivery_date: Optional[date] = None
    digital_signature: str | None = None
    fabrication_details: str | None = None
    budgeted_details: str | None = None
    pool_id: int | None = None
    pool_price: float = 0.0
    pool_currency: str = "ARS"
    pool_image: str | None = None
    pools_data: str | None = None
    adicionales_data: str | None = None
    design_observations: str | None = None
    important_observations: str | None = None
    notes: str | None = None
    # Optional per-work-order term overrides (JSON list, may be empty string).
    delivery_terms_override: str | None = ""
    warranty_override: str | None = ""
    date: Optional[datetime] = None


class WorkOrderCreate(WorkOrderBase):
    pass


class WorkOrderUpdate(BaseModel):
    status: str | None = None
    origin: str | None = None
    material: str | None = None
    material_price_m2: float | None = None
    materials_data: str | None = None
    color: str | None = None
    thickness: str | None = None
    finish: str | None = None
    bacha: str | None = None
    anafe: str | None = None
    currency: str | None = None
    usd_rate: float | None = None
    subtotal: float | None = None
    transport: float | None = None
    installation: float | None = None
    discount: float | None = None
    discount_percentage: float | None = None
    discount_fixed_amount: float | None = None
    total: float | None = None
    subtotal_usd: float | None = None
    transport_usd: float | None = None
    total_usd: float | None = None
    deposit_received: float | None = None
    deposit_currency: str | None = None
    deposit_usd: float | None = None
    balance_due: float | None = None
    balance_due_usd: float | None = None
    balance_paid: bool | None = None
    balance_paid_at: Optional[datetime] = None
    payment_method: str | None = None
    installments: int | None = None
    priority: str | None = None
    delivery_date: Optional[date] = None
    digital_signature: str | None = None
    signed_at: Optional[datetime] = None
    fabrication_details: str | None = None
    budgeted_details: str | None = None
    pool_id: int | None = None
    pool_price: float | None = None
    pool_currency: str | None = None
    pool_image: str | None = None
    stock_deducted: bool | None = None
    pools_data: str | None = None
    adicionales_data: str | None = None
    design_observations: str | None = None
    important_observations: str | None = None
    notes: str | None = None
    date: Optional[datetime] = None


class WorkOrderResponse(WorkOrderBase, BaseResponse):
    id: int
    number: str
    status: str
    origin: str
    stock_deducted: bool
    balance_paid: bool
    balance_paid_at: Optional[datetime] = None
    signed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_orm_with_client(cls, order) -> "WorkOrderResponse":
        """Build a response from a WorkOrder ORM row, populating client_*
        from the related Client (JOIN). The WorkOrder model has no
        client_name column — only client_id — so without this helper the
        response would have client_name=None."""
        data = cls.model_validate(order).model_dump()
        client = getattr(order, "client", None)
        if client:
            data["client_name"] = client.name or data.get("client_name") or ""
            data["client_phone"] = client.phone or data.get("client_phone") or ""
            data["client_email"] = client.email or data.get("client_email") or ""
            data["client_address"] = client.address or data.get("client_address") or ""
        return cls.model_validate(data)

